File: src/main/java/me/sargunvohra/leveluphp/core/ProgressionController.java
package me.sargunvohra.leveluphp.core;

import lombok.extern.log4j.Log4j2;
import lombok.val;
import me.sargunvohra.leveluphp.capability.PlayerLevelHandler;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.text.TextComponentString;
import net.minecraftforge.event.entity.living.LivingDeathEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;

/** This class is responsible for managing progression of the player's hp level. */
@Log4j2
public class ProgressionController {

  @SubscribeEvent
  void onLivingDeath(LivingDeathEvent event) {
    val source = event.getSource().getTrueSource();
    if (!(source instanceof EntityPlayer)) return;
    source.getCapability(PlayerLevelHandler.CAPABILITY).ifPresent(hpData -> {
      if (hpData.addXp(1)) source.sendMessage(new TextComponentString("Level up!"));
    });
  }
}

File: src/main/java/me/sargunvohra/leveluphp/core/ProgressionController.kt
package me.sargunvohra.leveluphp.core

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.text.TextComponentString
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.entity.living.LivingDeathEvent
import net.minecraftforge.eventbus.api.SubscribeEvent

/** This class is responsible for managing progression of the player's hp level.  */
class ProgressionController {

    @SubscribeEvent
    internal fun onLivingDeath(event: LivingDeathEvent) {
        val source = event.source.trueSource as? EntityPlayer ?: return
        source.playerLevelHandler.ifPresent {
            val oldLevel = it.level
            it.xp += 1
            if (it.level > oldLevel)
                source.sendMessage(TextComponentString("Level up!"))
        }
    }

    fun register() {
        MinecraftForge.EVENT_BUS.register(this)
    }
}
